### PR TITLE
Resource monitoring implementation

### PR DESCRIPTION
**Pull request by Kavaldrin**
PR without API changes
continuation of https://github.com/AliceO2Group/AliceO2/pull/3474

### CODE REVIEW

**Comment by ktf** [comment]
@awegrzyn what is the plan for the Monitoring release needed by this?

**Comment by ktf** [comment]
@Kavaldrin could you please resolve the conflicts and squash it to a single commit?

**Comment by awegrzyn** [comment]
@ktf There's PR https://github.com/AliceO2Group/Monitoring/pull/203 which requires rebasing, I'm happy to assist @Kavaldrin with this.

**Comment by ktf** [comment]
@awegrzyn thank you for your help!

File: Framework/Core/src/ResourcesMonitoringManager.cxx [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    if(line.length() < performanceMetricPrefix.length()) { return {}; }
+    
+    if(auto res = std::mismatch(performanceMetricPrefix.begin(), performanceMetricPrefix.end(), 
+        line.begin()); res.first != performanceMetricPrefix.end())
+    { return {}; }
+
+    std::istringstream iss(line);
+    std::vector<std::string> elements(std::istream_iterator<std::string>{iss}, std::istream_iterator<std::string>());
+
+    std::vector<PerformanceMetric> result(elements.size() - 4); // -1 prefix -1 main metric name -1 timestamp -1 hostname
```
Comment (ktf) [inline Framework/Core/src/ResourcesMonitoringManager.cxx:90]: The current parser is allocation-less. Please make sure it remains such. Why do you need a new one? Can't you adapt the old one?

Comment (Kavaldrin) [reply Framework/Core/src/ResourcesMonitoringManager.cxx:90]: The old parser works only for convention -> single metric - single value with type. Adam suggested to change it to -> single metric - multiple value which cannot be parsed with existing parser. I will change the new one to be allocation-less.

**Comment by ktf** [comment]
Can't you simply adapt the old parser? Why having a new one?

**Comment by Kavaldrin** [comment]
@ktf You mean to update old parser that will work with multiple values?

**Comment by ktf** [comment]
Yes, we have a parseMetrics, why introducing a specific one for the resource monitoring? Also, why do you need to send all the metrics as a multivalue field? can't you simply send them one after the other?

File: Framework/Core/src/ResourcesMonitoringManager.h [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+{
+public:
+
+    void enableResourcesMonitoring(unsigned short interval) noexcept
+        { m_enabled = true; m_interval = interval; }
+    void disableResourcesMonitoring() noexcept
+        { m_enabled = false; }
+
+    void addNewMetrics(const std::string& specName, std::vector<PerformanceMetric>&& parsedMetrics) noexcept;
+    bool dumpMetricsToJSON() noexcept;
```
Comment (ktf) [inline Framework/Core/src/ResourcesMonitoringManager.h:49]: Why do you need yet another metrics store? Can't you simply use the DeviceMetricInfos one?

**Comment by Kavaldrin** [comment]
@ktf Well one by one was in previous version. I'm a bit confused :^( 
There was discussion about it in monitoring PR https://github.com/AliceO2Group/Monitoring/pull/203

summon @awegrzyn

**Comment by awegrzyn** [comment]
Hi guys,
I see there's some confusion, could you lemme know whether these performance metrics need to picked up by filter? I though this is not the case.

The motivation of having multi-value metric is that in production the process monitoring will go to InfluxDB, sending each metric separately is very inefficient.

**Comment by Kavaldrin** [comment]
It seems that those parsed metrics are used only by debugGUI (which does not work for me :^(  ) or maybe I'm missing something.
@ktf What is your point of view about it?

**Comment by jotwinow** [comment]
@ktf @awegrzyn @Kavaldrin how do we proceed? We would like to finalize this project in June due to academic year constraints. As far as I know the monitoring should work in any environment (EPN, grid, local) and access to DB is not guaranteed.

**Comment by awegrzyn** [comment]
Hi,
As I've already mentioned, the parser only picks metrics which are then directed to DebugGUI, I don't think process monitoring shall be displayed in DebugGUI (@ktf Could you confirm?).
If that's the case please finish up Monitoring PR and then revert changes to parser as Giulio asked.

Cheers,
Adam

**Comment by ktf** [comment]
@awegrzyn no, that's not true. The metrics are sent to the driver process which accumulates them in the DeviceMetricInfos and **if you have a GUI** they get displayed there. They are accumulated nevertheless when you use the DPL driver. So the task should simply dump the contents of the the DeviceMetricInfos from the driver into some json format.

**Comment by Kavaldrin** [comment]
Well squash went not as I expected :^(

**Comment by awegrzyn** [comment]
Monitoring v3.2.0 release with additional process monitoring.

**Comment by jotwinow** [comment]
Ok. Thanks Adam.

**Comment by Kavaldrin** [comment]
new clean PR https://github.com/AliceO2Group/AliceO2/pull/3879

**Pull request closed without merging**